File: src/main/java/com/tkj/garagedoorz/config/GarageDoorzConfig.java
package com.tkj.garagedoorz.config;

import com.tkj.garagedoorz.domain.DoorStatus;
import com.tkj.garagedoorz.domain.GarageDoor;
import com.tkj.garagedoorz.domain.GarageDoorzHwController;
import com.tkj.garagedoorz.domain.HwController;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Profile;

import java.util.List;

import static java.util.Collections.singletonList;

@Configuration
public class GarageDoorzConfig {

    @Bean
    public HwController hwController() {

        return new HwController() {

            @Override
            public void pressDoorButton( int doorIndex ) {

            }

            @Override
            public boolean isDoorOpen( int doorIndex ) {

                return false;
            }

            @Override
            public List<DoorStatus> getGarageDoorStatuses() {

                return singletonList( new DoorStatus( "Test Door", false ) );
            }

        };
    }

    @Profile( "pi" )
    @Configuration
    public static class Pi {

        @Bean( "door1" )
        public GarageDoor door1(
                @Value( "${garagedoorz.doors.jill.label}" ) String doorName,
                @Value( "${garagedoorz.doors.jill.actuator}" ) int actuator,
                @Value( "${garagedoorz.doors.jill.position-sensor}" ) int positionSensor
        ) {

            return new GarageDoor( doorName, actuator, positionSensor );
        }

        @Bean( "door2" )
        public GarageDoor door2(
                @Value( "${garagedoorz.doors.tom.label}" ) String doorName,
                @Value( "${garagedoorz.doors.tom.actuator}" ) int actuator,
                @Value( "${garagedoorz.doors.tom.position-sensor}" ) int positionSensor
        ) {

            return new GarageDoor( doorName, actuator, positionSensor );
        }

        @Bean
        public HwController hwController( final GarageDoor door1, final GarageDoor door2 ) {

            return new GarageDoorzHwController( new GarageDoor[] { door1, door2 } );
        }

    }

}

File: src/main/java/com/tkj/garagedoorz/config/GarageDoorzConfig.kt
package com.tkj.garagedoorz.config

import com.pi4j.io.gpio.GpioController
import com.pi4j.io.gpio.GpioFactory
import com.tkj.garagedoorz.domain.GarageDoor
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Profile
import org.springframework.core.annotation.Order

@Configuration
@Profile( "pi" )
class GarageDoorzConfig {

    @Bean
    fun gpioController(): GpioController {

        return GpioFactory.getInstance()
    }

    @Bean( "door1" )
    @Order( 1 )
    fun door1(
            gpio: GpioController,
            @Value( "\${garagedoorz.doors.jill.label}" ) doorName: String,
            @Value( "\${garagedoorz.doors.jill.actuator}" ) actuator: Int,
            @Value( "\${garagedoorz.doors.jill.position-sensor}" ) positionSensor: Int
    ): GarageDoor {

        return GarageDoor( gpio, doorName, actuator, positionSensor )
    }

    @Bean( "door2" )
    @Order( 2 )
    fun door2(
            gpio: GpioController,
            @Value( "\${garagedoorz.doors.tom.label}" ) doorName: String,
            @Value( "\${garagedoorz.doors.tom.actuator}" ) actuator: Int,
            @Value( "\${garagedoorz.doors.tom.position-sensor}" ) positionSensor: Int
    ): GarageDoor {

        return GarageDoor( gpio, doorName, actuator, positionSensor )
    }

}
